vendor/bat-native-ledger/src/bat/ledger/internal/bitflyer/bitflyer.cc

```diff
@@ -0,0 +1,285 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+
+#include "base/guid.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/global_constants.h"
+#include "bat/ledger/internal/bitflyer/bitflyer.h"
+#include "bat/ledger/internal/bitflyer/bitflyer_authorization.h"
+#include "bat/ledger/internal/bitflyer/bitflyer_transfer.h"
+#include "bat/ledger/internal/bitflyer/bitflyer_util.h"
+#include "bat/ledger/internal/bitflyer/bitflyer_wallet.h"
+#include "bat/ledger/internal/common/time_util.h"
+#include "bat/ledger/internal/endpoint/bitflyer/bitflyer_server.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/logging/event_log_keys.h"
+#include "brave_base/random.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+namespace {
+const char kFeeMessage[] =
+    "5% transaction fee collected by Brave Software International";
+}  // namespace
+
+namespace ledger {
+namespace bitflyer {
+
+Bitflyer::Bitflyer(LedgerImpl* ledger)
+    : transfer_(std::make_unique<BitflyerTransfer>(ledger)),
+      authorization_(std::make_unique<BitflyerAuthorization>(ledger)),
+      wallet_(std::make_unique<BitflyerWallet>(ledger)),
+      bitflyer_server_(std::make_unique<endpoint::BitflyerServer>(ledger)),
+      ledger_(ledger) {}
+
+Bitflyer::~Bitflyer() = default;
+
+void Bitflyer::Initialize() {
+  auto wallet = GetWallet();
+  if (!wallet) {
+    return;
+  }
+
+  for (const auto& value : wallet->fees) {
+    StartTransferFeeTimer(value.first, 1);
+  }
+}
+
+void Bitflyer::StartContribution(const std::string& contribution_id,
+                                 type::ServerPublisherInfoPtr info,
+                                 const double amount,
+                                 ledger::ResultCallback callback) {
+  if (!info) {
+    BLOG(0, "Publisher info is null");
+    ContributionCompleted(type::Result::LEDGER_ERROR, "", contribution_id,
+                          amount, "", callback);
+    return;
+  }
+
+  const double fee = (amount * 1.05) - amount;
```

Comment:
nit: Isn't `amount * 0.05` better?

Reply:
I wonder if it was done this way to avoid some type of rounding error? But yeah, what you're suggesting certainly seems more straightforward.

Reply:
I agree this line is confusing - but since it's consistent with how Uphold fees are handled, I would probably prefer to leave it as-is for now. In the future, we'll want to take a look at moving the fee processing into a generic component, rather than duplicating it for each wallet provider; we can revisit then.